diff --git a/include/asm/offset.h b/include/asm/offset.h
index 6b7f128..b88fd26 100644
--- a/include/asm/offset.h
+++ b/include/asm/offset.h
@@ -29,21 +29,20 @@
 #define OFFSET_SIZE         (8*20)
 
 /* maincall used */ 
-#define OFFSET_UMAINCALL_T1     (8*0)
-#define OFFSET_UMAINCALL_T3     (8*1)
-#define OFFSET_UMAINCALL_RA     (8*2)
-#define OFFSET_UMAINCALL_A0     (8*3)
-#define OFFSET_UMAINCALL_A1     (8*4)
-#define OFFSET_UMAINCALL_A2     (8*5)
-#define OFFSET_UMAINCALL_A3     (8*6)
-#define OFFSET_UMAINCALL_A4     (8*7)
-#define OFFSET_UMAINCALL_A5     (8*8)
-#define OFFSET_UMAINCALL_A6     (8*9)
-#define OFFSET_UMAINCALL_A7     (8*10)
-#define OFFSET_UMAINCALL_SP     (8*11)
-
-
-
-#define OFFSET_UMAINCALL        (8*12)
+#define OFFSET_UMAINCALL_T0     (8*0)
+#define OFFSET_UMAINCALL_T1     (8*1)
+#define OFFSET_UMAINCALL_T3     (8*2)
+#define OFFSET_UMAINCALL_RA     (8*3)
+#define OFFSET_UMAINCALL_A0     (8*4)
+#define OFFSET_UMAINCALL_A1     (8*5)
+#define OFFSET_UMAINCALL_A2     (8*6)
+#define OFFSET_UMAINCALL_A3     (8*7)
+#define OFFSET_UMAINCALL_A4     (8*8)
+#define OFFSET_UMAINCALL_A5     (8*9)
+#define OFFSET_UMAINCALL_A6     (8*10)
+#define OFFSET_UMAINCALL_A7     (8*11)
+#define OFFSET_UMAINCALL_SP     (8*12)
+
+#define OFFSET_UMAINCALL        (8*13)
 
 #endif
\ No newline at end of file
diff --git a/include/umaincall.h b/include/umaincall.h
index 1a5e376..6ccaf50 100644
--- a/include/umaincall.h
+++ b/include/umaincall.h
@@ -25,6 +25,7 @@ typedef unsigned long reg_t;
 
 struct umaincall
 {
+    reg_t t0;
     reg_t t1;
     reg_t t3;
     reg_t ra;
@@ -40,18 +41,19 @@ struct umaincall
 };
 
 
-#define OFFSET_UMAINCALL_T1     (8*0)
-#define OFFSET_UMAINCALL_T3     (8*1)
-#define OFFSET_UMAINCALL_RA     (8*2)
-#define OFFSET_UMAINCALL_A0     (8*3)
-#define OFFSET_UMAINCALL_A1     (8*4)
-#define OFFSET_UMAINCALL_A2     (8*5)
-#define OFFSET_UMAINCALL_A3     (8*6)
-#define OFFSET_UMAINCALL_A4     (8*7)
-#define OFFSET_UMAINCALL_A5     (8*8)
-#define OFFSET_UMAINCALL_A6     (8*9)
-#define OFFSET_UMAINCALL_A7     (8*10)
-#define OFFSET_UMAINCALL_SP     (8*11)
+#define OFFSET_UMAINCALL_T0     (8*0)
+#define OFFSET_UMAINCALL_T1     (8*1)
+#define OFFSET_UMAINCALL_T3     (8*2)
+#define OFFSET_UMAINCALL_RA     (8*3)
+#define OFFSET_UMAINCALL_A0     (8*4)
+#define OFFSET_UMAINCALL_A1     (8*5)
+#define OFFSET_UMAINCALL_A2     (8*6)
+#define OFFSET_UMAINCALL_A3     (8*7)
+#define OFFSET_UMAINCALL_A4     (8*8)
+#define OFFSET_UMAINCALL_A5     (8*9)
+#define OFFSET_UMAINCALL_A6     (8*10)
+#define OFFSET_UMAINCALL_A7     (8*11)
+#define OFFSET_UMAINCALL_SP     (8*12)
 
 
 #define OFFSET_UMAINCALL        sizeof(struct umaincall)
diff --git a/src/maincall/umaincall_entry.S b/src/maincall/umaincall_entry.S
index 43144bb..32417f9 100644
--- a/src/maincall/umaincall_entry.S
+++ b/src/maincall/umaincall_entry.S
@@ -4,7 +4,7 @@
 // MAINCALL used registers
 .macro SAVE_MAINCALL
     addi sp, sp,  -OFFSET_UMAINCALL
-
+    sd t0, OFFSET_UMAINCALL_T0(sp)
     sd t1, OFFSET_UMAINCALL_T1(sp)
     sd t3, OFFSET_UMAINCALL_T3(sp)
     sd ra, OFFSET_UMAINCALL_RA(sp)
@@ -18,11 +18,11 @@
     sd a7, OFFSET_UMAINCALL_A7(sp)
     addi t0, sp, OFFSET_UMAINCALL
     sd t0, OFFSET_UMAINCALL_SP(sp)
-    
 .endm
 
 
 .macro RESTORE_MAINCALL
+    ld t0, OFFSET_UMAINCALL_T0(sp)
     ld t1, OFFSET_UMAINCALL_T1(sp)
     ld t3, OFFSET_UMAINCALL_T3(sp)
     ld ra, OFFSET_UMAINCALL_RA(sp)
@@ -34,8 +34,7 @@
     ld a5, OFFSET_UMAINCALL_A5(sp)
     ld a6, OFFSET_UMAINCALL_A6(sp)
     ld a7, OFFSET_UMAINCALL_A7(sp)
-
-    addi sp, sp,  OFFSET_UMAINCALL
+    ld sp, OFFSET_UMAINCALL_SP(sp)
 .endm
 
 
diff --git a/src/start/stage3.c b/src/start/stage3.c
index cb09e1f..fbb8555 100644
--- a/src/start/stage3.c
+++ b/src/start/stage3.c
@@ -51,10 +51,10 @@ void _dasics_entry_stage3(uint64_t sp, rtld_fini fini)
     original_libcfg_alloc(DASICS_LIBCFG_V | DASICS_LIBCFG_R, \
                         (uint64_t)_umain_elf_table->got_begin, \
                         (uint64_t)_umain_elf_table->got_begin + sizeof(uint64_t) * (_umain_elf_table->got_num + 2));
-    original_libcfg_alloc(DASICS_LIBCFG_V | DASICS_LIBCFG_R | DASICS_LIBCFG_W, \
-                0, \
-                TASK_SIZE);
-    original_jumpcfg_alloc(0, TASK_SIZE);
+    // original_libcfg_alloc(DASICS_LIBCFG_V | DASICS_LIBCFG_R | DASICS_LIBCFG_W, \
+    //             0, \
+    //             TASK_SIZE);
+    // original_jumpcfg_alloc(0, TASK_SIZE);
 
     // setup user ufault handler 
     csr_write(0x005, (uint64_t)dasics_ufault_entry);
