diff --git a/auto/cc/conf b/auto/cc/conf
index ba31cb88c..e6d753c7f 100644
--- a/auto/cc/conf
+++ b/auto/cc/conf
@@ -183,7 +183,7 @@ if [ "$NGX_PLATFORM" != win32 ]; then
     else
         ngx_feature="gcc builtin atomic operations"
         ngx_feature_name=NGX_HAVE_GCC_ATOMIC
-        ngx_feature_run=yes
+        ngx_feature_run=no
         ngx_feature_incs=
         ngx_feature_path=
         ngx_feature_libs=
@@ -204,7 +204,7 @@ if [ "$NGX_PLATFORM" != win32 ]; then
     else
         ngx_feature="C99 variadic macros"
         ngx_feature_name="NGX_HAVE_C99_VARIADIC_MACROS"
-        ngx_feature_run=yes
+        ngx_feature_run=no
         ngx_feature_incs="#include <stdio.h>
 #define var(dummy, ...)  sprintf(__VA_ARGS__)"
         ngx_feature_path=
@@ -218,7 +218,7 @@ if [ "$NGX_PLATFORM" != win32 ]; then
 
     ngx_feature="gcc variadic macros"
     ngx_feature_name="NGX_HAVE_GCC_VARIADIC_MACROS"
-    ngx_feature_run=yes
+    ngx_feature_run=no
     ngx_feature_incs="#include <stdio.h>
 #define var(dummy, args...)  sprintf(args)"
     ngx_feature_path=
diff --git a/auto/cc/name b/auto/cc/name
index ded93f5bc..d6ab27a6a 100644
--- a/auto/cc/name
+++ b/auto/cc/name
@@ -7,7 +7,7 @@ if [ "$NGX_PLATFORM" != win32 ]; then
 
     ngx_feature="C compiler"
     ngx_feature_name=
-    ngx_feature_run=yes
+    ngx_feature_run=no
     ngx_feature_incs=
     ngx_feature_path=
     ngx_feature_libs=
diff --git a/auto/lib/libatomic/conf b/auto/lib/libatomic/conf
index 8c8cb438b..fe6e924cd 100644
--- a/auto/lib/libatomic/conf
+++ b/auto/lib/libatomic/conf
@@ -6,9 +6,11 @@
 if [ $NGX_LIBATOMIC != YES ]; then
 
     have=NGX_HAVE_LIBATOMIC . auto/have
-    CORE_INCS="$CORE_INCS $NGX_LIBATOMIC/src"
-    LINK_DEPS="$LINK_DEPS $NGX_LIBATOMIC/src/libatomic_ops.a"
-    CORE_LIBS="$CORE_LIBS $NGX_LIBATOMIC/src/libatomic_ops.a"
+    CORE_INCS="$CORE_INCS $NGX_LIBATOMIC/include"
+    LINK_DEPS="$LINK_DEPS $NGX_LIBATOMIC/lib/libatomic_ops.a"
+    CORE_LIBS="$CORE_LIBS $NGX_LIBATOMIC/lib/libatomic_ops.a"
+    NGX_LIBATOMIC_ROOTFS=$NGX_LIBATOMIC
+    NGX_LIBATOMIC=ROOTFS
 
 else
 
diff --git a/auto/lib/make b/auto/lib/make
index b64e32908..24d48570a 100644
--- a/auto/lib/make
+++ b/auto/lib/make
@@ -3,19 +3,19 @@
 # Copyright (C) Nginx, Inc.
 
 
-if [ $PCRE != NONE -a $PCRE != NO -a $PCRE != YES ]; then
+if [ $PCRE != NONE -a $PCRE != NO -a $PCRE != YES -a $PCRE != ROOTFS ]; then
     . auto/lib/pcre/make
 fi
 
-if [ $OPENSSL != NONE -a $OPENSSL != NO -a $OPENSSL != YES ]; then
+if [ $OPENSSL != NONE -a $OPENSSL != NO -a $OPENSSL != YES -a $OPENSSL != ROOTFS ]; then
     . auto/lib/openssl/make
 fi
 
-if [ $ZLIB != NONE -a $ZLIB != NO -a $ZLIB != YES ]; then
+if [ $ZLIB != NONE -a $ZLIB != NO -a $ZLIB != YES -a $ZLIB != ROOTFS ]; then
     . auto/lib/zlib/make
 fi
 
-if [ $NGX_LIBATOMIC != NO -a $NGX_LIBATOMIC != YES ]; then
+if [ $NGX_LIBATOMIC != NO -a $NGX_LIBATOMIC != YES -a $NGX_LIBATOMIC != ROOTFS ]; then
     . auto/lib/libatomic/make
 fi
 
diff --git a/auto/lib/openssl/conf b/auto/lib/openssl/conf
index fdf430dff..761483b50 100644
--- a/auto/lib/openssl/conf
+++ b/auto/lib/openssl/conf
@@ -40,12 +40,15 @@ if [ $OPENSSL != NONE ]; then
         ;;
 
         *)
-            CORE_INCS="$CORE_INCS $OPENSSL/.openssl/include"
-            CORE_DEPS="$CORE_DEPS $OPENSSL/.openssl/include/openssl/ssl.h"
-            CORE_LIBS="$CORE_LIBS $OPENSSL/.openssl/lib/libssl.a"
-            CORE_LIBS="$CORE_LIBS $OPENSSL/.openssl/lib/libcrypto.a"
+            CORE_INCS="$CORE_INCS $OPENSSL/include"
+            CORE_DEPS="$CORE_DEPS $OPENSSL/include/openssl/ssl.h"
+            # CORE_LIBS="$CORE_LIBS $OPENSSL/lib/libssl.a"
+            # CORE_LIBS="$CORE_LIBS $OPENSSL/lib/libcrypto.a"
+            CORE_LIBS="$CORE_LIBS -L $OPENSSL/lib -lcrypto -lssl"
             CORE_LIBS="$CORE_LIBS $NGX_LIBDL"
             CORE_LIBS="$CORE_LIBS $NGX_LIBPTHREAD"
+            OPENSSL_ROOTFS=$OPENSSL
+            OPENSSL=ROOTFS
 
             if [ "$NGX_PLATFORM" = win32 ]; then
                 CORE_LIBS="$CORE_LIBS -lgdi32 -lcrypt32 -lws2_32"
diff --git a/auto/lib/pcre/conf b/auto/lib/pcre/conf
index cdf1809f5..c73d70857 100644
--- a/auto/lib/pcre/conf
+++ b/auto/lib/pcre/conf
@@ -5,7 +5,7 @@
 
 if [ $PCRE != NONE ]; then
 
-    if [ -f $PCRE/src/pcre2.h.generic ]; then
+    if [ -f $PCRE/include/pcre2.h ]; then
 
         PCRE_LIBRARY=PCRE2
 
@@ -16,8 +16,8 @@ if [ $PCRE != NONE ]; then
             have=PCRE2_STATIC . auto/have
         fi
 
-        CORE_INCS="$CORE_INCS $PCRE/src/"
-        CORE_DEPS="$CORE_DEPS $PCRE/src/pcre2.h"
+        CORE_INCS="$CORE_INCS $PCRE/include/"
+        CORE_DEPS="$CORE_DEPS $PCRE/include/pcre2.h"
 
         case "$NGX_CC_NAME" in
 
@@ -27,8 +27,11 @@ if [ $PCRE != NONE ]; then
             ;;
 
             *)
-                LINK_DEPS="$LINK_DEPS $PCRE/.libs/libpcre2-8.a"
-                CORE_LIBS="$CORE_LIBS $PCRE/.libs/libpcre2-8.a"
+                #LINK_DEPS="$LINK_DEPS $PCRE/.libs/libpcre2-8.a"
+                #CORE_LIBS="$CORE_LIBS $PCRE/.libs/libpcre2-8.a"
+                CORE_LIBS="$CORE_LIBS -L $PCRE/lib -lpcre2-8"
+                PCRE_ROOTFS=$PCRE
+                PCRE=ROOTFS
             ;;
 
         esac
@@ -43,8 +46,8 @@ if [ $PCRE != NONE ]; then
             have=PCRE_STATIC . auto/have
         fi
 
-        CORE_INCS="$CORE_INCS $PCRE"
-        CORE_DEPS="$CORE_DEPS $PCRE/pcre.h"
+        CORE_INCS="$CORE_INCS $PCRE/include"
+        CORE_DEPS="$CORE_DEPS $PCRE/include/pcre.h"
 
         case "$NGX_CC_NAME" in
 
@@ -54,8 +57,11 @@ if [ $PCRE != NONE ]; then
             ;;
 
             *)
-                LINK_DEPS="$LINK_DEPS $PCRE/.libs/libpcre.a"
-                CORE_LIBS="$CORE_LIBS $PCRE/.libs/libpcre.a"
+                # LINK_DEPS="$LINK_DEPS $PCRE/lib/libpcre.a"
+                # CORE_LIBS="$CORE_LIBS $PCRE/lib/libpcre.a"
+                CORE_LIBS="$CORE_LIBS -L $PCRE/lib -lpcre"
+                PCRE_ROOTFS=$PCRE
+                PCRE=ROOTFS
             ;;
 
         esac
diff --git a/auto/lib/zlib/conf b/auto/lib/zlib/conf
index 239592e3c..a32388abf 100644
--- a/auto/lib/zlib/conf
+++ b/auto/lib/zlib/conf
@@ -4,7 +4,7 @@
 
 
 if [ $ZLIB != NONE ]; then
-    CORE_INCS="$CORE_INCS $ZLIB"
+    CORE_INCS="$CORE_INCS $ZLIB/include"
 
     case "$NGX_CC_NAME" in
 
@@ -33,9 +33,11 @@ if [ $ZLIB != NONE ]; then
 
         *)
             have=NGX_ZLIB . auto/have
-            LINK_DEPS="$LINK_DEPS $ZLIB/libz.a"
-            CORE_LIBS="$CORE_LIBS $ZLIB/libz.a"
-            #CORE_LIBS="$CORE_LIBS -L $ZLIB -lz"
+            # LINK_DEPS="$LINK_DEPS $ZLIB/lib/libz.a"
+            # CORE_LIBS="$CORE_LIBS $ZLIB/lib/libz.a"
+            CORE_LIBS="$CORE_LIBS -L $ZLIB/lib -lz"
+            ZLIB_ROOTFS=$ZLIB
+            ZLIB=ROOTFS
         ;;
 
     esac
diff --git a/auto/os/linux b/auto/os/linux
index bc0556b3a..05af6a003 100644
--- a/auto/os/linux
+++ b/auto/os/linux
@@ -36,7 +36,7 @@ fi
 
 ngx_feature="epoll"
 ngx_feature_name="NGX_HAVE_EPOLL"
-ngx_feature_run=yes
+ngx_feature_run=no
 ngx_feature_incs="#include <sys/epoll.h>"
 ngx_feature_path=
 ngx_feature_libs=
@@ -135,7 +135,7 @@ ngx_feature_test="int fd; struct stat sb;
 CC_AUX_FLAGS="$cc_aux_flags -D_GNU_SOURCE"
 ngx_feature="sendfile()"
 ngx_feature_name="NGX_HAVE_SENDFILE"
-ngx_feature_run=yes
+ngx_feature_run=no
 ngx_feature_incs="#include <sys/sendfile.h>
                   #include <errno.h>"
 ngx_feature_path=
@@ -156,7 +156,7 @@ fi
 CC_AUX_FLAGS="$cc_aux_flags -D_GNU_SOURCE -D_FILE_OFFSET_BITS=64"
 ngx_feature="sendfile64()"
 ngx_feature_name="NGX_HAVE_SENDFILE64"
-ngx_feature_run=yes
+ngx_feature_run=no
 ngx_feature_incs="#include <sys/sendfile.h>
                   #include <errno.h>"
 ngx_feature_path=
@@ -174,7 +174,7 @@ ngx_include="sys/prctl.h"; . auto/include
 
 ngx_feature="prctl(PR_SET_DUMPABLE)"
 ngx_feature_name="NGX_HAVE_PR_SET_DUMPABLE"
-ngx_feature_run=yes
+ngx_feature_run=no
 ngx_feature_incs="#include <sys/prctl.h>"
 ngx_feature_path=
 ngx_feature_libs=
@@ -186,7 +186,7 @@ ngx_feature_test="if (prctl(PR_SET_DUMPABLE, 1, 0, 0, 0) == -1) return 1"
 
 ngx_feature="prctl(PR_SET_KEEPCAPS)"
 ngx_feature_name="NGX_HAVE_PR_SET_KEEPCAPS"
-ngx_feature_run=yes
+ngx_feature_run=no
 ngx_feature_incs="#include <sys/prctl.h>"
 ngx_feature_path=
 ngx_feature_libs=
diff --git a/auto/summary b/auto/summary
index b3c07eedc..28e646ea3 100644
--- a/auto/summary
+++ b/auto/summary
@@ -18,6 +18,7 @@ else
     case $PCRE in
         YES)   echo "  + using system $PCRE_LIBRARY library" ;;
         NONE)  echo "  + PCRE library is not used" ;;
+        ROOTFS)echo "  + using $PCRE_LIBRARY library from rootfs: $PCRE_ROOTFS" ;;
         *)     echo "  + using $PCRE_LIBRARY library: $PCRE" ;;
     esac
 fi
@@ -25,12 +26,14 @@ fi
 case $OPENSSL in
     YES)   echo "  + using system OpenSSL library" ;;
     NONE)  echo "  + OpenSSL library is not used" ;;
+    ROOTFS)echo "  + using OpenSSL library from rootfs: $OPENSSL_ROOTFS" ;;
     *)     echo "  + using OpenSSL library: $OPENSSL" ;;
 esac
 
 case $ZLIB in
     YES)   echo "  + using system zlib library" ;;
     NONE)  echo "  + zlib library is not used" ;;
+    ROOTFS)echo "  + using zlib library from rootfs: $ZLIB_ROOTFS" ;;
     *)     echo "  + using zlib library: $ZLIB" ;;
 esac
 
diff --git a/auto/types/sizeof b/auto/types/sizeof
index 480d8cfa4..52c7287a8 100644
--- a/auto/types/sizeof
+++ b/auto/types/sizeof
@@ -33,7 +33,7 @@ int main(void) {
 END
 
 
-ngx_test="$CC $CC_TEST_FLAGS $CC_AUX_FLAGS \
+ngx_test="gcc $CC_TEST_FLAGS $CC_AUX_FLAGS \
           -o $NGX_AUTOTEST $NGX_AUTOTEST.c $NGX_LD_OPT $ngx_feature_libs"
 
 eval "$ngx_test >> $NGX_AUTOCONF_ERR 2>&1"
diff --git a/auto/unix b/auto/unix
index f29e69c61..da0106c19 100644
--- a/auto/unix
+++ b/auto/unix
@@ -99,7 +99,7 @@ if test -z "$NGX_KQUEUE_CHECKED"; then
 
         ngx_feature="kqueue's EVFILT_TIMER"
         ngx_feature_name="NGX_HAVE_TIMER_EVENT"
-        ngx_feature_run=yes
+        ngx_feature_run=no
         ngx_feature_incs="#include <sys/event.h>
                           #include <sys/time.h>"
         ngx_feature_path=
@@ -853,7 +853,7 @@ ngx_feature_test="void *p; p = memalign(4096, 4096);
 
 ngx_feature="mmap(MAP_ANON|MAP_SHARED)"
 ngx_feature_name="NGX_HAVE_MAP_ANON"
-ngx_feature_run=yes
+ngx_feature_run=no
 ngx_feature_incs="#include <sys/mman.h>"
 ngx_feature_path=
 ngx_feature_libs=
@@ -866,7 +866,7 @@ ngx_feature_test="void *p;
 
 ngx_feature='mmap("/dev/zero", MAP_SHARED)'
 ngx_feature_name="NGX_HAVE_MAP_DEVZERO"
-ngx_feature_run=yes
+ngx_feature_run=no
 ngx_feature_incs="#include <sys/mman.h>
                   #include <sys/stat.h>
                   #include <fcntl.h>"
@@ -881,7 +881,7 @@ ngx_feature_test='void *p; int  fd;
 
 ngx_feature="System V shared memory"
 ngx_feature_name="NGX_HAVE_SYSVSHM"
-ngx_feature_run=yes
+ngx_feature_run=no
 ngx_feature_incs="#include <sys/ipc.h>
                   #include <sys/shm.h>"
 ngx_feature_path=
@@ -895,7 +895,7 @@ ngx_feature_test="int  id;
 
 ngx_feature="POSIX semaphores"
 ngx_feature_name="NGX_HAVE_POSIX_SEM"
-ngx_feature_run=yes
+ngx_feature_run=no
 ngx_feature_incs="#include <semaphore.h>"
 ngx_feature_path=
 ngx_feature_libs=
